File: src/teslausb/gadget.py
"""USB mass storage gadget management.

This module manages the Linux USB gadget subsystem to present disk images
as USB mass storage devices to the Tesla vehicle.

The gadget is configured via configfs at /sys/kernel/config/usb_gadget/.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

logger = logging.getLogger(__name__)


class GadgetError(Exception):
    """Error during gadget operations."""


@dataclass
class LunConfig:
    """Configuration for a single LUN (logical unit).

    Args:
        disk_path: Path to the disk image file
        removable: Whether the device appears as removable media
        readonly: Whether to expose as read-only
        cdrom: Whether to emulate a CD-ROM drive
    """

    disk_path: Path
    removable: bool = True
    readonly: bool = False
    cdrom: bool = False


class Gadget(Protocol):
    """Protocol for USB gadget operations."""

    def initialize(self, luns: dict[int, LunConfig]) -> None:
        """Create and configure the gadget."""
        ...

    def remove(self) -> None:
        """Remove the gadget configuration."""
        ...

    def enable(self) -> None:
        """Bind gadget to UDC - host sees the drives."""
        ...

    def disable(self) -> None:
        """Unbind gadget from UDC - drives disappear."""
        ...

    def is_enabled(self) -> bool:
        """Check if gadget is currently bound to UDC."""
        ...

    def is_initialized(self) -> bool:
        """Check if gadget structure exists."""
        ...


class UsbGadget:
    """USB mass storage gadget using Linux configfs.

    This class manages the USB gadget subsystem to present disk images
    as USB mass storage devices. It requires:
    - configfs mounted at /sys/kernel/config
    - libcomposite kernel module loaded
    - dwc2 (or similar) USB gadget controller

    Example:
        gadget = UsbGadget()
        gadget.initialize({
            0: LunConfig(disk_path=Path("/backingfiles/cam_disk.bin")),
        })
        gadget.enable()
    """

    # USB IDs for Linux Foundation composite gadget
    VENDOR_ID = "0x1d6b"
    PRODUCT_ID = "0x0104"

    def __init__(
        self,
        name: str = "teslausb",
        configfs: Path = Path("/sys/kernel/config/usb_gadget"),
    ):
        """Initialize UsbGadget.

        Args:
            name: Gadget name (creates directory under configfs)
            configfs: Path to USB gadget configfs mount point
        """
        self.name = name
        self.configfs = configfs
        self.path = configfs / name
        self._udc_path = Path("/sys/class/udc")

    def _write(self, path: Path, value: str) -> None:
        """Write value to a configfs file."""
        logger.debug(f"Writing '{value}' to {path}")
        path.write_text(value)

    def _read(self, path: Path) -> str:
        """Read value from a configfs file."""
        return path.read_text().strip()

    def _get_udc(self) -> str:
        """Get available USB Device Controller name.

        Returns:
            Name of the first available UDC

        Raises:
            GadgetError: If no UDC is available
        """
        if not self._udc_path.exists():
            raise GadgetError(f"UDC path {self._udc_path} does not exist")

        udcs = list(self._udc_path.iterdir())
        if not udcs:
            raise GadgetError("No USB Device Controller found")

        return udcs[0].name

    def initialize(self, luns: dict[int, LunConfig]) -> None:
        """Create gadget structure in configfs.

        Args:
            luns: Mapping of LUN number to configuration.
                  LUN 0 is typically the camera disk, LUN 1 is music, etc.

        Raises:
            GadgetError: If initialization fails
        """
        if self.is_initialized():
            logger.info(f"Gadget {self.name} already initialized")
            return

        if not luns:
            raise GadgetError("At least one LUN must be configured")

        # Check prerequisites
        if not self.configfs.exists():
            configfs_base = Path("/sys/kernel/config")
            if not configfs_base.exists():
                raise GadgetError(
                    f"configfs not mounted. "
                    "Run: sudo mount -t configfs none /sys/kernel/config"
                )
            else:
                raise GadgetError(
                    f"USB gadget configfs not available at {self.configfs}. "
                    "Run: sudo modprobe libcomposite"
                )

        # Check that cam_disk exists
        for lun_id, config in luns.items():
            if not config.disk_path.exists():
                raise GadgetError(f"Disk image not found: {config.disk_path}")

        logger.info(f"Initializing gadget {self.name} with {len(luns)} LUN(s)")

        try:
            # Create gadget directory
            self.path.mkdir(parents=True)

            # Set USB IDs
            self._write(self.path / "idVendor", self.VENDOR_ID)
            self._write(self.path / "idProduct", self.PRODUCT_ID)
            self._write(self.path / "bcdDevice", "0x0100")
            self._write(self.path / "bcdUSB", "0x0200")

            # Create strings (English)
            strings = self.path / "strings" / "0x409"
            strings.mkdir(parents=True)
            self._write(strings / "manufacturer", "TeslaUSB")
            self._write(strings / "product", "Tesla USB Drive")
            self._write(strings / "serialnumber", "fedcba9876543210")

            # Create mass storage function
            func = self.path / "functions" / "mass_storage.0"
            func.mkdir(parents=True)

            # Configure each LUN
            for lun_id, config in sorted(luns.items()):
                self._configure_lun(func, lun_id, config)

            # Create configuration
            cfg = self.path / "configs" / "c.1"
            cfg.mkdir(parents=True)

            cfg_strings = cfg / "strings" / "0x409"
            cfg_strings.mkdir(parents=True)
            self._write(cfg_strings / "configuration", "Config 1: Mass Storage")

            self._write(cfg / "MaxPower", "250")

            # Link function to configuration
            link = cfg / "mass_storage.0"
            if not link.exists():
                link.symlink_to(func)

            logger.info(f"Gadget {self.name} initialized")

        except OSError as e:
            logger.error(f"Failed to initialize gadget: {e}")
            # Try to clean up partial initialization
            self._cleanup_partial()
            raise GadgetError(f"Failed to initialize gadget: {e}") from e

    def _configure_lun(self, func_path: Path, lun_id: int, config: LunConfig) -> None:
        """Configure a single LUN.

        Args:
            func_path: Path to the mass_storage function
            lun_id: LUN number (0, 1, 2, ...)
            config: LUN configuration
        """
        lun = func_path / f"lun.{lun_id}"

        # LUN 0 is created automatically, others need mkdir
        if lun_id > 0:
            lun.mkdir(exist_ok=True)

        logger.debug(f"Configuring LUN {lun_id}: {config.disk_path}")

        self._write(lun / "removable", "1" if config.removable else "0")
        self._write(lun / "ro", "1" if config.readonly else "0")
        self._write(lun / "cdrom", "1" if config.cdrom else "0")
        self._write(lun / "file", str(config.disk_path))

    def _cleanup_partial(self) -> None:
        """Clean up a partial gadget setup after failure."""
        try:
            if self.path.exists():
                import shutil

                shutil.rmtree(self.path)
        except OSError:
            pass

    def remove(self) -> None:
        """Remove gadget from configfs.

        This will disable the gadget first if it's enabled.
        """
        if not self.is_initialized():
            return

        logger.info(f"Removing gadget {self.name}")

        # Must disable first
        self.disable()

        try:
            # Remove symlink from config
            cfg_link = self.path / "configs" / "c.1" / "mass_storage.0"
            if cfg_link.is_symlink():
                cfg_link.unlink()

            # Remove config strings
            cfg_strings = self.path / "configs" / "c.1" / "strings" / "0x409"
            if cfg_strings.exists():
                cfg_strings.rmdir()

            # Remove config
            cfg = self.path / "configs" / "c.1"
            if cfg.exists():
                cfg.rmdir()

            # Remove LUNs (except lun.0 which is automatic)
            func = self.path / "functions" / "mass_storage.0"
            if func.exists():
                for lun in func.iterdir():
                    if lun.is_dir() and lun.name != "lun.0":
                        lun.rmdir()
                func.rmdir()

            # Remove strings
            strings = self.path / "strings" / "0x409"
            if strings.exists():
                strings.rmdir()

            # Remove gadget
            self.path.rmdir()

            logger.info(f"Gadget {self.name} removed")

        except OSError as e:
            logger.error(f"Failed to remove gadget: {e}")
            raise GadgetError(f"Failed to remove gadget: {e}") from e

    def enable(self) -> None:
        """Bind gadget to UDC - Tesla sees the drives.

        Raises:
            GadgetError: If gadget is not initialized or binding fails
        """
        if not self.is_initialized():
            raise GadgetError("Gadget not initialized")

        if self.is_enabled():
            logger.debug("Gadget already enabled")
            return

        udc = self._get_udc()
        logger.info(f"Enabling gadget {self.name} on {udc}")

        try:
            self._write(self.path / "UDC", udc)
        except OSError as e:
            raise GadgetError(f"Failed to enable gadget: {e}") from e

    def disable(self) -> None:
        """Unbind gadget from UDC - drives disappear from Tesla."""
        if not self.is_enabled():
            return

        logger.info(f"Disabling gadget {self.name}")

        try:
            self._write(self.path / "UDC", "")
        except OSError as e:
            logger.warning(f"Failed to disable gadget: {e}")

    def is_enabled(self) -> bool:
        """Check if gadget is bound to UDC.

        Returns:
            True if gadget is bound and visible to host
        """
        udc_file = self.path / "UDC"
        if not udc_file.exists():
            return False
        return bool(self._read(udc_file))

    def is_initialized(self) -> bool:
        """Check if gadget structure exists in configfs.

        Returns:
            True if gadget directory exists
        """
        return self.path.exists()

    def get_status(self) -> dict:
        """Get current gadget status.

        Returns:
            Dictionary with status information
        """
        status = {
            "name": self.name,
            "initialized": self.is_initialized(),
            "enabled": False,
            "udc": None,
            "luns": {},
        }

        if not self.is_initialized():
            return status

        # Get UDC
        udc_file = self.path / "UDC"
        if udc_file.exists():
            udc = self._read(udc_file)
            status["enabled"] = bool(udc)
            status["udc"] = udc if udc else None

        # Get LUN info
        func = self.path / "functions" / "mass_storage.0"
        if func.exists():
            for lun_dir in sorted(func.iterdir()):
                if lun_dir.is_dir() and lun_dir.name.startswith("lun."):
                    lun_id = int(lun_dir.name.split(".")[1])
                    file_path = lun_dir / "file"
                    status["luns"][lun_id] = {
                        "file": self._read(file_path) if file_path.exists() else None,
                        "readonly": self._read(lun_dir / "ro") == "1",
                    }

        return status


class MockGadget:
    """Mock gadget for testing.

    Tracks state in memory without touching the filesystem.
    """

    def __init__(self):
        self._initialized = False
        self._enabled = False
        self.luns: dict[int, LunConfig] = {}
        self.enable_count = 0
        self.disable_count = 0

    def initialize(self, luns: dict[int, LunConfig]) -> None:
        """Initialize mock gadget."""
        if not luns:
            raise GadgetError("At least one LUN must be configured")
        self.luns = {k: v for k, v in luns.items()}
        self._initialized = True

    def remove(self) -> None:
        """Remove mock gadget."""
        self._enabled = False
        self._initialized = False
        self.luns.clear()

    def enable(self) -> None:
        """Enable mock gadget."""
        if not self._initialized:
            raise GadgetError("Gadget not initialized")
        self._enabled = True
        self.enable_count += 1

    def disable(self) -> None:
        """Disable mock gadget."""
        if self._enabled:
            self._enabled = False
            self.disable_count += 1

    def is_enabled(self) -> bool:
        """Check if mock gadget is enabled."""
        return self._enabled

    def is_initialized(self) -> bool:
        """Check if mock gadget is initialized."""
        return self._initialized

    def get_status(self) -> dict:
        """Get mock gadget status."""
        return {
            "name": "mock",
            "initialized": self._initialized,
            "enabled": self._enabled,
            "udc": "mock-udc" if self._enabled else None,
            "luns": {
                k: {"file": str(v.disk_path), "readonly": v.readonly}
                for k, v in self.luns.items()
            },
        }

File: src/gadget.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

use crate::command::CommandRunner;
use crate::error::{Error, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LunConfig {
    pub disk_path: PathBuf,
    pub removable: bool,
    pub readonly: bool,
    pub cdrom: bool,
}

impl LunConfig {
    pub fn new(disk_path: PathBuf) -> Self {
        Self {
            disk_path,
            removable: true,
            readonly: false,
            cdrom: false,
        }
    }
}

#[derive(Debug, Clone)]
pub struct UsbGadget {
    name: String,
    configfs: PathBuf,
    path: PathBuf,
    udc_path: PathBuf,
}

impl Default for UsbGadget {
    fn default() -> Self {
        Self::new("teslausb", PathBuf::from("/sys/kernel/config/usb_gadget"))
    }
}

impl UsbGadget {
    const VENDOR_ID: &'static str = "0x1d6b";
    const PRODUCT_ID: &'static str = "0x0104";

    pub fn new(name: impl Into<String>, configfs: PathBuf) -> Self {
        let name = name.into();
        let path = configfs.join(&name);
        Self {
            name,
            configfs,
            path,
            udc_path: PathBuf::from("/sys/class/udc"),
        }
    }

    pub fn initialize(&self, luns: &[(u32, LunConfig)]) -> Result<()> {
        if self.is_initialized() {
            return Ok(());
        }
        if luns.is_empty() {
            return Err(Error::new("at least one LUN must be configured"));
        }

        if !self.configfs.exists() {
            let configfs_base = Path::new("/sys/kernel/config");
            if !configfs_base.exists() {
                return Err(Error::new(
                    "configfs is not mounted; run: sudo mount -t configfs none /sys/kernel/config",
                ));
            }
            if !run_modprobe("libcomposite") || !self.configfs.exists() {
                return Err(Error::new(
                    "USB gadget configfs is unavailable; run: sudo modprobe libcomposite",
                ));
            }
        }

        for (_, lun) in luns {
            if !lun.disk_path.exists() {
                return Err(Error::new(format!(
                    "disk image not found: {}",
                    lun.disk_path.display()
                )));
            }
        }

        if let Err(err) = self.initialize_inner(luns) {
            let _ = fs::remove_dir_all(&self.path);
            return Err(err);
        }
        Ok(())
    }

    fn initialize_inner(&self, luns: &[(u32, LunConfig)]) -> Result<()> {
        fs::create_dir_all(&self.path)?;
        self.write(&self.path.join("idVendor"), Self::VENDOR_ID)?;
        self.write(&self.path.join("idProduct"), Self::PRODUCT_ID)?;
        self.write(&self.path.join("bcdDevice"), "0x0100")?;
        self.write(&self.path.join("bcdUSB"), "0x0200")?;

        let strings = self.path.join("strings/0x409");
        fs::create_dir_all(&strings)?;
        self.write(&strings.join("manufacturer"), "TeslaUSB")?;
        self.write(&strings.join("product"), "Tesla USB Drive")?;
        self.write(&strings.join("serialnumber"), "fedcba9876543210")?;

        let func = self.path.join("functions/mass_storage.0");
        fs::create_dir_all(&func)?;
        for (lun_id, config) in luns {
            self.configure_lun(&func, *lun_id, config)?;
        }

        let cfg = self.path.join("configs/c.1");
        fs::create_dir_all(cfg.join("strings/0x409"))?;
        self.write(
            &cfg.join("strings/0x409/configuration"),
            "Config 1: Mass Storage",
        )?;
        self.write(&cfg.join("MaxPower"), "250")?;

        let link = cfg.join("mass_storage.0");
        if !link.exists() {
            std::os::unix::fs::symlink(&func, &link)?;
        }
        Ok(())
    }

    fn configure_lun(&self, func: &Path, lun_id: u32, config: &LunConfig) -> Result<()> {
        let lun = func.join(format!("lun.{lun_id}"));
        fs::create_dir_all(&lun)?;
        self.write(
            &lun.join("removable"),
            if config.removable { "1" } else { "0" },
        )?;
        self.write(&lun.join("ro"), if config.readonly { "1" } else { "0" })?;
        self.write(&lun.join("cdrom"), if config.cdrom { "1" } else { "0" })?;
        self.write(&lun.join("file"), &config.disk_path.display().to_string())?;
        Ok(())
    }

    pub fn remove(&self) -> Result<()> {
        if !self.is_initialized() {
            return Ok(());
        }
        self.disable()?;

        let cfg_link = self.path.join("configs/c.1/mass_storage.0");
        if cfg_link.is_symlink() {
            fs::remove_file(cfg_link)?;
        }
        remove_dir_if_exists(&self.path.join("configs/c.1/strings/0x409"))?;
        remove_dir_if_exists(&self.path.join("configs/c.1"))?;

        let func = self.path.join("functions/mass_storage.0");
        if func.exists() {
            for entry in fs::read_dir(&func)? {
                let entry = entry?;
                let path = entry.path();
                if path.is_dir() && entry.file_name().to_string_lossy() != "lun.0" {
                    remove_dir_if_exists(&path)?;
                }
            }
            remove_dir_if_exists(&func)?;
        }

        remove_dir_if_exists(&self.path.join("strings/0x409"))?;
        remove_dir_if_exists(&self.path)?;
        Ok(())
    }

    pub fn enable(&self) -> Result<()> {
        if !self.is_initialized() {
            return Err(Error::new("gadget is not initialized"));
        }
        if self.is_enabled() {
            return Ok(());
        }
        let udc = self.get_udc()?;
        self.write(&self.path.join("UDC"), &udc)?;
        Ok(())
    }

    pub fn disable(&self) -> Result<()> {
        if !self.is_enabled() {
            return Ok(());
        }
        self.write(&self.path.join("UDC"), "")?;
        Ok(())
    }

    pub fn is_initialized(&self) -> bool {
        self.path.exists()
    }

    pub fn is_enabled(&self) -> bool {
        let udc = self.path.join("UDC");
        udc.exists()
            && fs::read_to_string(udc)
                .map(|content| !content.trim().is_empty())
                .unwrap_or(false)
    }

    pub fn status_json(&self) -> String {
        let enabled = self.is_enabled();
        let initialized = self.is_initialized();
        let udc = if initialized {
            fs::read_to_string(self.path.join("UDC"))
                .unwrap_or_default()
                .trim()
                .to_string()
        } else {
            String::new()
        };
        format!(
            "{{\n  \"name\": \"{}\",\n  \"initialized\": {},\n  \"enabled\": {},\n  \"udc\": {}\n}}\n",
            self.name,
            initialized,
            enabled,
            if udc.is_empty() {
                "null".to_string()
            } else {
                format!("\"{}\"", udc)
            }
        )
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    fn write(&self, path: &Path, value: &str) -> Result<()> {
        fs::write(path, format!("{value}\n"))?;
        Ok(())
    }

    fn get_udc(&self) -> Result<String> {
        if !self.udc_path.exists() {
            return Err(Error::new(format!(
                "UDC path {} does not exist",
                self.udc_path.display()
            )));
        }
        let mut udcs = fs::read_dir(&self.udc_path)?
            .filter_map(|entry| entry.ok())
            .map(|entry| entry.file_name().to_string_lossy().to_string())
            .collect::<Vec<_>>();
        udcs.sort();
        udcs.into_iter()
            .next()
            .ok_or_else(|| Error::new("no USB Device Controller found"))
    }
}

#[derive(Debug)]
pub struct GadgetDisableGuard {
    gadget: UsbGadget,
    was_enabled: bool,
}

impl GadgetDisableGuard {
    pub fn disable_if_needed(gadget: UsbGadget) -> Result<Self> {
        let was_enabled = gadget.is_enabled();
        if was_enabled {
            gadget.disable()?;
            if gadget.is_enabled() {
                return Err(Error::new("gadget is still enabled after disable"));
            }
        }
        Ok(Self {
            gadget,
            was_enabled,
        })
    }
}

impl Drop for GadgetDisableGuard {
    fn drop(&mut self) {
        if self.was_enabled {
            if let Err(err) = self.gadget.enable() {
                eprintln!("error: failed to re-enable USB gadget: {}", err);
            }
        }
    }
}

fn run_modprobe(module: &str) -> bool {
    CommandRunner
        .run("modprobe", [module], Some(Duration::from_secs(30)))
        .map(|output| output.success())
        .unwrap_or(false)
}

fn remove_dir_if_exists(path: &Path) -> Result<()> {
    if path.exists() {
        fs::remove_dir(path)?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use std::time::{SystemTime, UNIX_EPOCH};

    use super::*;

    fn temp_dir(name: &str) -> PathBuf {
        let suffix = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let path = std::env::temp_dir().join(format!("teslausb-{name}-{suffix}"));
        fs::create_dir_all(&path).unwrap();
        path
    }

    #[test]
    fn lun_config_defaults_match_mass_storage_expectations() {
        let config = LunConfig::new(PathBuf::from("/cam.bin"));
        assert_eq!(config.disk_path, PathBuf::from("/cam.bin"));
        assert!(config.removable);
        assert!(!config.readonly);
        assert!(!config.cdrom);
    }

    #[test]
    fn initialized_and_enabled_reflect_configfs_files() {
        let configfs = temp_dir("configfs");
        let gadget = UsbGadget::new("test", configfs.clone());
        assert_eq!(gadget.name(), "test");
        assert!(!gadget.is_initialized());
        assert!(!gadget.is_enabled());

        fs::create_dir_all(configfs.join("test")).unwrap();
        fs::write(configfs.join("test/UDC"), "").unwrap();
        assert!(gadget.is_initialized());
        assert!(!gadget.is_enabled());

        fs::write(configfs.join("test/UDC"), "fake-udc\n").unwrap();
        assert!(gadget.is_enabled());

        let _ = fs::remove_dir_all(configfs);
    }

    #[test]
    fn initialize_rejects_empty_luns_and_missing_disk() {
        let configfs = temp_dir("configfs");
        let gadget = UsbGadget::new("test", configfs.clone());

        assert!(gadget.initialize(&[]).is_err());
        assert!(gadget
            .initialize(&[(0, LunConfig::new(PathBuf::from("/definitely/missing.img")))])
            .is_err());

        let _ = fs::remove_dir_all(configfs);
    }

    #[test]
    fn initialize_writes_expected_configfs_structure() {
        let root = temp_dir("gadget-init");
        let configfs = root.join("usb_gadget");
        fs::create_dir_all(&configfs).unwrap();
        let disk = root.join("cam.bin");
        fs::write(&disk, b"cam").unwrap();

        let gadget = UsbGadget::new("test", configfs.clone());
        gadget
            .initialize(&[(0, LunConfig::new(disk.clone()))])
            .unwrap();

        let gadget_path = configfs.join("test");
        assert_eq!(
            fs::read_to_string(gadget_path.join("idVendor")).unwrap(),
            "0x1d6b\n"
        );
        assert_eq!(
            fs::read_to_string(gadget_path.join("functions/mass_storage.0/lun.0/file")).unwrap(),
            format!("{}\n", disk.display())
        );
        assert!(gadget_path.join("configs/c.1/mass_storage.0").is_symlink());

        let _ = fs::remove_dir_all(root);
    }

    #[test]
    fn enable_requires_initialized_gadget_and_udc() {
        let root = temp_dir("gadget-enable");
        let configfs = root.join("usb_gadget");
        let udc_path = root.join("udc");
        fs::create_dir_all(&configfs).unwrap();
        fs::create_dir_all(&udc_path).unwrap();
        fs::create_dir_all(udc_path.join("fake-udc")).unwrap();
        fs::create_dir_all(configfs.join("test")).unwrap();
        fs::write(configfs.join("test/UDC"), "").unwrap();

        let mut gadget = UsbGadget::new("test", configfs.clone());
        gadget.udc_path = udc_path;
        gadget.enable().unwrap();
        assert_eq!(
            fs::read_to_string(configfs.join("test/UDC")).unwrap(),
            "fake-udc\n"
        );
        gadget.disable().unwrap();
        assert!(!gadget.is_enabled());

        let _ = fs::remove_dir_all(root);
    }

    #[test]
    fn enable_fails_when_not_initialized_or_no_udc_exists() {
        let root = temp_dir("gadget-enable-fail");
        let configfs = root.join("usb_gadget");
        let udc_path = root.join("udc");
        fs::create_dir_all(&configfs).unwrap();
        fs::create_dir_all(&udc_path).unwrap();

        let mut gadget = UsbGadget::new("test", configfs.clone());
        gadget.udc_path = udc_path.clone();
        assert!(gadget.enable().is_err());

        fs::create_dir_all(configfs.join("test")).unwrap();
        fs::write(configfs.join("test/UDC"), "").unwrap();
        assert!(gadget.enable().is_err());

        let _ = fs::remove_dir_all(root);
    }

    #[test]
    fn status_json_reports_basic_state() {
        let configfs = temp_dir("gadget-status");
        fs::create_dir_all(configfs.join("test")).unwrap();
        fs::write(configfs.join("test/UDC"), "fake\n").unwrap();
        let gadget = UsbGadget::new("test", configfs.clone());

        let json = gadget.status_json();

        assert!(json.contains("\"name\": \"test\""));
        assert!(json.contains("\"initialized\": true"));
        assert!(json.contains("\"enabled\": true"));
        assert!(json.contains("\"udc\": \"fake\""));

        let _ = fs::remove_dir_all(configfs);
    }

    #[test]
    fn disable_guard_reenables_when_it_disabled_the_gadget() {
        let root = temp_dir("gadget-guard");
        let configfs = root.join("usb_gadget");
        let udc_path = root.join("udc");
        fs::create_dir_all(configfs.join("test")).unwrap();
        fs::write(configfs.join("test/UDC"), "fake\n").unwrap();
        fs::create_dir_all(udc_path.join("fake")).unwrap();

        let mut gadget = UsbGadget::new("test", configfs.clone());
        gadget.udc_path = udc_path;
        {
            let _guard = GadgetDisableGuard::disable_if_needed(gadget.clone()).unwrap();
            assert!(!gadget.is_enabled());
        }
        assert!(gadget.is_enabled());

        let _ = fs::remove_dir_all(root);
    }
}
